// @mongosh/shell-evaluator/lib/shell-evaluator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShellEvaluator = void 0;
const shell_api_1 = require("@mongosh/shell-api");
const async_rewriter2_1 = __importDefault(require("@mongosh/async-rewriter2"));
const history_1 = require("@mongosh/history");
let hasAlreadyRunGlobalRuntimeSupportEval = false;
let v8;
try {
    v8 = require('v8');
}
catch (_c) {
}
if ((_b = (_a = v8 === null || v8 === void 0 ? void 0 : v8.startupSnapshot) === null || _a === void 0 ? void 0 : _a.isBuildingSnapshot) === null || _b === void 0 ? void 0 : _b.call(_a)) {
    v8.startupSnapshot.addSerializeCallback(() => {
        eval(new async_rewriter2_1.default().runtimeSupportCode());
        hasAlreadyRunGlobalRuntimeSupportEval = true;
    });
}
class ShellEvaluator {
    constructor(instanceState, resultHandler = shell_api_1.toShellResult) {
        this.hasAppliedAsyncWriterRuntimeSupport = true;
        this.instanceState = instanceState;
        this.resultHandler = resultHandler;
        this.asyncWriter = new async_rewriter2_1.default();
        this.hasAppliedAsyncWriterRuntimeSupport = false;
    }
    async innerEval(originalEval, input, context, filename) {
        var _a, _b, _c, _d, _e;
        const { shellApi } = this.instanceState;
        const trimmedInput = input.trim();
        const argv = trimmedInput.replace(/;$/, '').split(/\s+/g);
        const cmd = argv.shift();
        if (((_a = shellApi[cmd]) === null || _a === void 0 ? void 0 : _a.isDirectShellCommand) &&
            ((_b = shellApi[cmd]) === null || _b === void 0 ? void 0 : _b.acceptsRawInput) &&
            !((_c = argv[0]) !== null && _c !== void 0 ? _c : '').startsWith('(')) {
            const rawArg = trimmedInput.replace(/^\S+\s*/, '');
            return shellApi[cmd](rawArg);
        }
        if (((_d = shellApi[cmd]) === null || _d === void 0 ? void 0 : _d.isDirectShellCommand) &&
            !((_e = argv[0]) !== null && _e !== void 0 ? _e : '').startsWith('(')) {
            return shellApi[cmd](...argv);
        }
        let rewrittenInput = this.asyncWriter.process(input);
        const hiddenCommands = RegExp(history_1.HIDDEN_COMMANDS, 'g');
        if (!hiddenCommands.test(input) && !hiddenCommands.test(rewrittenInput)) {
            this.instanceState.messageBus.emit('mongosh:evaluate-input', {
                input: (0, history_1.redactSensitiveData)(trimmedInput),
            });
        }
        if (!this.hasAppliedAsyncWriterRuntimeSupport) {
            this.hasAppliedAsyncWriterRuntimeSupport = true;
            const supportCode = this.asyncWriter.runtimeSupportCode();
            if (!hasAlreadyRunGlobalRuntimeSupportEval) {
                eval(supportCode);
            }
            rewrittenInput = supportCode + ';\n' + rewrittenInput;
        }
        try {
            return await originalEval(rewrittenInput, context, filename);
        }
        catch (err) {
            throw this.instanceState.transformError(err);
        }
    }
    async customEval(originalEval, input, context, filename) {
        const evaluationResult = await this.innerEval(originalEval, input, context, filename);
        return await this.resultHandler(evaluationResult);
    }
}
exports.ShellEvaluator = ShellEvaluator;
